lib/route.js

```diff
@@ -154,43 +155,61 @@ function buildRouting (options) {
-        afterRouteAdded.call(this, { path: path.slice(1) }, notHandledErr, done)
-      } else {
-        afterRouteAdded.call(this, { path }, notHandledErr, done)
+    const path = opts.url || opts.path
+    if (path === '/' && prefix.length > 0 && opts.method !== 'HEAD') {
+      switch (opts.prefixTrailingSlash) {
+        case 'slash':
+          addNewRoute.call(this, path)
+          break
+        case 'no-slash':
+          addNewRoute.call(this, '')
+          break
+        case 'both':
+        default:
+          addNewRoute.call(this, '')
+          // If ignoreTrailingSlash is set to true we need to add only the '' route to prevent adding an incomplete one.
+          if (ignoreTrailingSlash !== true) {
+            addNewRoute.call(this, path, true)
+          }
       }
-    })
+    } else if (path[0] === '/' && prefix.endsWith('/')) {
+      // Ensure that '/prefix/' + '/route' gets registered as '/prefix/route'
+      addNewRoute.call(this, path.slice(1))
+    } else {
+      addNewRoute.call(this, path)
+    }
+
+    // TODO: remove once review is done
+    // this.after((notHandledErr, done) => {
+    //   const path = opts.url || opts.path
+    //   if (path === '/' && prefix.length > 0 && opts.method !== 'HEAD') {
+    //     switch (opts.prefixTrailingSlash) {
+    //       case 'slash':
+    //         afterRouteAdded.call(this, { path }, notHandledErr, done)
+    //         break
+    //       case 'no-slash':
+    //         afterRouteAdded.call(this, { path: '' }, notHandledErr, done)
+    //         break
+    //       case 'both':
+    //       default:
+    //         afterRouteAdded.call(this, { path: '' }, notHandledErr, done)
+    //         // If ignoreTrailingSlash is set to true we need to add only the '' route to prevent adding an incomplete one.
+    //         if (ignoreTrailingSlash !== true) {
+    //           afterRouteAdded.call(this, { path, prefixing: true }, notHandledErr, done)
+    //         }
+    //     }
+    //   } else if (path[0] === '/' && prefix.endsWith('/')) {
+    //     // Ensure that '/prefix/' + '/route' gets registered as '/prefix/route'
+    //     afterRouteAdded.call(this, { path: path.slice(1) }, notHandledErr, done)
+    //   } else {
+    //     afterRouteAdded.call(this, { path }, notHandledErr, done)
+    //   }
+    // })
 
     // chainable api
     return this
 
-    /**
-     * This function sets up a new route, its log serializers, and triggers route hooks.
-     *
-     * @param {object} opts contains route `path` and `prefixing` flag which indicates if this is an auto-prefixed route, e.g. `fastify.register(routes, { prefix: '/foo' })`
-     * @param {*} notHandledErr error object to be passed back to the original invoker
-     * @param {*} done callback
-     */
-    function afterRouteAdded ({ path, prefixing = false }, notHandledErr, done) {
+    function addNewRoute (path, prefixing = false) {
```

Comment:
I would try to move this to an arrow function to avoid calling it with `this`.

Reply:
Sure, but that would make the function unreachable and will need to either declare it outside of `buildRouting` or in someplace within [route.js:L43-L45](https://github.com/fastify/fastify/blob/next/lib/route.js#L43-L45)

Or there's some other way to do it and I'm missing it?

Reply:
@mcollina is your concern addressed?